main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala

```diff
@@ -137,14 +134,159 @@ object SafetyError {
+
+    override def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> The following supercast attempts to cast a Java type to a non-Java type.
+         |
+         |${code(loc, "the supercast occurs here.")}
+         |
+         |Java type:        ${formatJavaType(javaType)}
+         |Tried casting to: $nonJavaType
          |""".stripMargin
     }
 
     override def explain(formatter: Formatter): Option[String] = None
   }
 
+  /**
+    * An error raised to indicate use of supercast from a type variable,
+    * i.e. the actual type is not known, possibly caused by
+    * supercasting a type to a type also being supercast.
+    *
+    * @param tvar the type of the expression being supercast (in this case a type variable).
+    * @param to   the type being cast to, i.e. the type of the supercast expression itself.
+    * @param loc  the source location of the supercast.
+    */
+  case class FromTypeVariableSupercast(tvar: Type, to: Type, loc: SourceLocation) extends SafetyError {
+    override def summary: String = "Unsafe supercast: Attempted to cast from a type variable."
+
+    override def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> The following supercast attempts to cast from a type variable.
+         |
+         |${code(loc, "the supercast occurs here.")}
+         |
+         |Type variable:    $tvar
+         |Tried casting to: $to
+         |""".stripMargin
+    }
+
+    override def explain(formatter: Formatter): Option[String] =
+      Some({
+        s"""Did you try to supercast two types at the same time? e.g.
+           |
+           |  if (true) supercast(a) else supercast(b)
+           |
+           |where a and b are Java types
+           |""".stripMargin
+      })
+  }
+
+  /**
+    * An error raised to indicate use of supercast to a type variable,
+    * i.e. the expected type is not known, possibly caused by
+    * supercasting a type to a type also being supercast.
+    *
+    * @param from the type of the expression being supercast.
+    * @param tvar the type being cast to, i.e. the type of the supercast expression itself (in this case a type variable).
+    * @param loc  the source location of the supercast.
+    */
+  case class ToTypeVariableSupercast(from: Type, tvar: Type, loc: SourceLocation) extends SafetyError {
+    override def summary: String = "Unsafe supercast: Attempted to cast to a type variable."
+
+    override def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> The following supercast attempts to cast to a type variable.
+         |
+         |${code(loc, "the supercast occurs here.")}
+         |
+         |Actual type:      $from
+         |Tried casting to: $tvar
+         |""".stripMargin
+    }
+
+    override def explain(formatter: Formatter): Option[String] =
```

Comment:
This also commonly occurs when we try to supercast the final value of a function, right?

```scala
def f(): ##java.lang.Object = supercast("hello")
```